flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/TestFlinkUpsert.java

```diff
@@ -165,6 +165,52 @@ public void testUpsertAndQuery() {
     }
   }
 
+  @Test
+  public void testUpsertOptions() {
+    String tableName = "test_upsert_options";
+    LocalDate dt20220301 = LocalDate.of(2022, 3, 1);
+    LocalDate dt20220302 = LocalDate.of(2022, 3, 2);
+
+    tableUpsertProps.remove(TableProperties.UPSERT_ENABLED);
+    sql(
+        "CREATE TABLE %s(id INT NOT NULL, province STRING NOT NULL, dt DATE, PRIMARY KEY(id,province) NOT ENFORCED) "
+            + "PARTITIONED BY (province) WITH %s",
+        tableName, toWithClause(tableUpsertProps));
+
+    try {
+      sql(
+          "INSERT INTO %s /*+ OPTIONS('upsert-enabled'='true')*/  VALUES "
+              + "(1, 'a', DATE '2022-03-01'),"
+              + "(2, 'b', DATE '2022-03-01'),"
+              + "(1, 'b', DATE '2022-03-01')",
+          tableName);
+
+      sql(
+          "INSERT INTO %s /*+ OPTIONS('upsert-enabled'='true')*/  VALUES "
+              + "(4, 'a', DATE '2022-03-02'),"
+              + "(5, 'b', DATE '2022-03-02'),"
+              + "(1, 'b', DATE '2022-03-02')",
+          tableName);
+
+      List<Row> rowsOn20220301 =
+          Lists.newArrayList(Row.of(2, "b", dt20220301), Row.of(1, "a", dt20220301));
+      TestHelpers.assertRows(
```

Comment:
not very familiar with upsert. for my own learning, why are the two queries resulted in different output? the primary key is the combination/concatenation of (id, province).

Reply:
In UPSERT mode, data with the same primary key is overwritten by new data. In this case, after two insertion operations, the data in the table is:
```
(1, 'a', DATE '2022-03-01')
(2, 'b', DATE '2022-03-01')
(4, 'a', DATE '2022-03-02')
(5, 'b', DATE '2022-03-02')
(1, 'b', DATE '2022-03-02')
```

Reply:
I realized where my confusion came from. I assumed the table is partitioned by date. Hence couldn't get my head around why the upsert (composite primary key) didn't work for day 1.

Reply:
We should consider refactoring these in a follow up PR so the table is partitioned by date.

I've encountered this confusion as well before.

Reply:
Non-blocking though =)

Reply:
😂  It's easy to misunderstand.